spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/groovy/org/springframework/cloud/contract/verifier/plugin/GradleContractsDownloader.groovy

```diff
@@ -26,73 +25,69 @@ class GradleContractsDownloader {
+								deleteStubsAfterTest, contractsProperties, failOnNoContracts).createNewInclusionProperties(cachedFolder)
+
+				return new DownloadedData(
+						downloadedContracts: contractsSubDirIfPresent(cachedFolder, logger),
+						inclusionProperties: inclusionProperties
+				)
 			}
-			File downloadedContracts = contractDownloader(extension, configuration).unpackedDownloadedContracts(config)
-			downloadedContract.put(configuration, downloadedContracts)
-			return downloadedContracts
 		}
-		this.log.info("For project [{}] will use contracts provided in the folder [{}]", this.project.name, defaultContractsDir)
-		return defaultContractsDir
-	}
+		final ContractDownloader contractDownloader =
+				contractDownloader(configuration, contractRepository, contractsPath, contractsMode, deleteStubsAfterTest,
+						contractsProperties, failOnNoContracts);
+		final File downloadedContracts = contractDownloader.unpackAndDownloadContracts();
+		final ContractDownloader.InclusionProperties inclusionProperties =
+				contractDownloader.createNewInclusionProperties(downloadedContracts)
 
-	private boolean shouldDownloadContracts(ContractVerifierExtension extension) {
-		return StringUtils.hasText(extension.contractDependency.getArtifactId()) ||
-				StringUtils.hasText(extension.contractDependency.getStringNotation()) ||
-				StringUtils.hasText(extension.contractRepository.repositoryUrl)
-	}
+		downloadedContract.put(configuration, downloadedContracts)
 
-	protected ContractDownloader contractDownloader(ContractVerifierExtension extension, StubConfiguration configuration) {
-		return new ContractDownloader(stubDownloader(extension), configuration,
-				extension.contractsPath, this.project.group as String, this.project.name, this.project.version as String)
+		return new DownloadedData(
+				downloadedContracts: contractsSubDirIfPresent(downloadedContracts, logger),
+				inclusionProperties: inclusionProperties
+		)
 	}
 
-	protected StubDownloader stubDownloader(ContractVerifierExtension extension) {
-		StubDownloaderBuilderProvider provider = new StubDownloaderBuilderProvider()
-		return provider.get(options(extension))
-	}
-
-	protected StubRunnerOptions options(ContractVerifierExtension extension) {
-		StubRunnerOptionsBuilder options = new StubRunnerOptionsBuilder()
-				.withOptions(StubRunnerOptions.fromSystemProps())
-				.withStubRepositoryRoot(extension.contractRepository.repositoryUrl)
-				.withStubsMode(extension.contractsMode)
-				.withUsername(extension.contractRepository.username)
-				.withPassword(extension.contractRepository.password)
-				.withDeleteStubsAfterTest(extension.deleteStubsAfterTest)
-				.withProperties(extension.contractsProperties)
-		if (extension.contractRepository.proxyPort) {
-			options = options.withProxy(extension.contractRepository.proxyHost, extension.contractRepository.proxyPort)
+	static File contractsSubDirIfPresent(File contractsDirectory, Logger logger) {
```

Comment:
I would think this wouldn't need to be static and probably should be private.

Reply:
Used in the `SpringCloudContractVerifierGradlePlugin`

Reply:
Ahh, I see now. So something to note about this, is that usually you want the outputs to be a directory within the build folder since:

1) They're gitignored, so changing the data won't result in cascading files into a commit where that doesn't make sense.
2) Further cleans up the input/output relationship to the task. 
2a) You shouldn't need to let the task know about the directory conventions once you're using the lazy configuration API. Essentially, the extension could define the intermediate output directory that contracts should be placed in before being processed by the generation tasks. That output directory would then be the input to the generation tasks.

Reply:
So, the logic here is:
1. We have an `init` task (used to be part of the `copy contracts`, but I extracted it to the dedicated one to make it incremental). That task is:
1.1 If configured, downloading and unpacking contracts into some temporary directory, which is then used as an input by `copy contracts` task. In this case, that output directory is calculated dynamically by the task itself and then we pick subdirectory from there using this `contractsSubDirIfPresent` method;
1.2 If downloading is not needed - just using default `contracts` directory within `resources` as an input for `copy contracts` task. Again, `contractsSubDirIfPresent` is used there to pick subdirectory if present. That default directory is kind of static and initialised in the `SpringCloudContractVerifierGradlePlugin` as a default output;
2. Copy contracts task is taking that directory from `init` task and copying those contracts to the `build` folder.

So it kind of does what you suggested, but not in the `init` task. Most of the times that `init` task does nothing, actually.

Reply:
I actually moved those properties to the internal nested config class and clean this part up a bit.